types/wlr_input_method_v2.c

```diff
@@ -108,10 +116,139 @@ static void im_delete_surrounding_text(struct wl_client *client,
 	input_method->pending.delete.after_length = after_length;
 }
 
+struct wlr_input_popup_surface_v2 *wlr_input_popup_surface_v2_from_wlr_surface(
+		struct wlr_surface *surface) {
+	assert(wlr_surface_is_input_popup_surface_v2(surface));
+	return (struct wlr_input_popup_surface_v2 *)surface->role_data;
+}
+
+void wlr_input_popup_surface_v2_send_text_input_rectangle(
+		struct wlr_input_popup_surface_v2 *popup_surface,
+		struct wlr_box *sbox) {
+	zwp_input_popup_surface_v2_send_text_input_rectangle(
+		popup_surface->resource, sbox->x, sbox->y, sbox->width, sbox->height);
+}
+
+static void popup_surface_set_mapped(
+		struct wlr_input_popup_surface_v2 *popup_surface, bool mapped) {
+	if (mapped && !popup_surface->mapped) {
+		popup_surface->mapped = true;
+		wlr_signal_emit_safe(&popup_surface->events.map, popup_surface);
+	} else if (!mapped && popup_surface->mapped) {
+		wlr_signal_emit_safe(&popup_surface->events.unmap, popup_surface);
+		popup_surface->mapped = false;
+	}
+}
+
+static void popup_surface_surface_role_commit(struct wlr_surface *surface) {
+	struct wlr_input_popup_surface_v2 *popup_surface = surface->role_data;
+	if (popup_surface == NULL) {
+		return;
+	}
+	popup_surface_set_mapped(popup_surface, wlr_surface_has_buffer(surface)
+		&& popup_surface->input_method->client_active);
+}
+
+static const struct wlr_surface_role input_popup_surface_v2_role = {
+	.name = "zwp_input_popup_surface_v2",
+	.commit = popup_surface_surface_role_commit,
+};
+
+bool wlr_surface_is_input_popup_surface_v2(struct wlr_surface *surface) {
+	return surface->role == &input_popup_surface_v2_role;
+}
+
+static void popup_surface_destroy(
+		struct wlr_input_popup_surface_v2 *popup_surface) {
+	popup_surface_set_mapped(popup_surface, false);
+	wlr_signal_emit_safe(&popup_surface->events.destroy, NULL);
+	wl_list_remove(&popup_surface->surface_destroy.link);
+	wl_list_remove(&popup_surface->link);
+	wl_resource_set_user_data(popup_surface->resource, NULL);
+	free(popup_surface);
+}
+
+static void popup_surface_handle_surface_destroy(struct wl_listener *listener,
+		void *data) {
+	struct wlr_input_popup_surface_v2 *popup_surface =
+		wl_container_of(listener, popup_surface, surface_destroy);
+	popup_surface_destroy(popup_surface);
+}
+
+static void popup_resource_destroy(struct wl_resource *resource) {
+	struct wlr_input_popup_surface_v2 *popup_surface =
+		wl_resource_get_user_data(resource);
+	if (popup_surface == NULL) {
+		return;
+	}
+	popup_surface_destroy(popup_surface);
+}
+
+static void popup_destroy(struct wl_client *client,
+		struct wl_resource *resource) {
+	struct wlr_input_popup_surface_v2 *popup_surface =
+		wl_resource_get_user_data(resource);
+	if (popup_surface == NULL) {
+		return;
+	}
+	popup_surface_destroy(popup_surface);
```

Comment:
`destroy` requests should simply call `wl_resource_destroy`. They just exist to inform the compositor that the client will no longer use the object.